File: src/main/java/org/fountainmc/api/command/AbstractCommand.java
package org.fountainmc.api.command;

/**
 * Represents a command.
 * Alternative to using annotations.
 */
public abstract class AbstractCommand<T extends CommandSender> {
    public abstract String name();

    public String permission() {
        return "";
    }

    public String noPermissionMessage() {
        return "You do not have permission to run this command";
    }

    public String usage() {
        return "";
    }

    public String[] aliases() {
        return new String[0];
    }

    public String description() {
        return "";
    }

    public abstract Class<T> allow();

    public abstract void onExecute(T sender, String[] args);

}

File: src/main/kotlin/org/fountainmc/api/command/AbstractCommand.kt
package org.fountainmc.api.command

import com.google.common.collect.ImmutableList
import org.fountainmc.api.internal.utils.emptyImmutableList

/**
 * Represents a command.
 * Alternative to using annotations.
 */
abstract class AbstractCommand<in T : CommandSender> {

    abstract fun name(): String

    val permission: String?
        get() = null

    val noPermissionMessage
        get() = "You do not have permission to run this command"

    val usage: String?
        get() = null

    val aliases: ImmutableList<String>
        get() = emptyImmutableList()

    val description: String?
        get() = null

    abstract fun allow(): Class<@UnsafeVariance T>

    abstract fun onExecute(sender: T, args: Array<String>)

}
